### PR TITLE
Fourier surface geometry features for tau_y/z gap (RFF input lift)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

High-frequency Fourier features for surface geometry will improve wall shear τ_y/τ_z prediction by giving the model the ability to detect sharp geometric transitions (leading edges, door gaps, wheel arches) that simple sinusoidal positional encodings cannot capture.

**Motivation:** The two dominant remaining gaps are τ_y (2.35× AB-UPT) and τ_z (2.73× AB-UPT). These errors are concentrated at sharp curvature regions where flow separates and reattaches — exactly where low-frequency spatial encodings fail. Random Fourier Features (RFF, Rahimi & Recht 2007) with high-frequency bands sample the spectrum of the surface geometry at frequencies the current `--learnable-pe` STRING-sep encoding misses. Composing RFF-augmented geometry inputs with the existing STRING-sep PE should give the attention heads finer-grained spatial structure to localise shear components.

**Prior art:** Neural Tangent Kernel studies (Tancik et al. 2020, "Fourier Features Let Networks Learn High-Frequency Functions in Low Dimensional Domains") showed that RFF input mappings dramatically improve convergence at high spatial frequency for 3D shape regression tasks. The DrivAerML surface has a known spectral gap: curvature features (κ₁, κ₂, merged PR #580) helped τ_y/z but only at the lowest curvature deciles; sharper edges likely need a richer frequency basis.

**Mechanism:** Apply a learnable random Fourier feature (RFF) map to the surface XYZ coordinates before feeding them into the model, concatenating the lifted features to the existing surface input. The mapping is `γ(x) = [sin(B·x), cos(B·x)]` where B is a (d_rff × 3) matrix of frequencies drawn from N(0, σ²). Both σ (frequency scale) and optionally B itself can be made learnable. This adds ~2·d_rff channels to the surface input and costs negligible FLOPS.

## Instructions

Add a command-line flag `--surface-rff-dim` (default=0 → disabled) and `--surface-rff-sigma` (default=10.0) to `train.py`. When `--surface-rff-dim D` is set:

1. At model init, allocate a fixed buffer `B` of shape `(D, 3)` drawn from `N(0, σ²)` (where σ = `--surface-rff-sigma`). Make `B` a non-trainable `nn.Parameter` with `requires_grad=False`, or store as `register_buffer`. This way it is saved in checkpoints without being optimised.
2. Before the surface features are fed into the Transolver encoder, lift the surface XYZ coordinates through: `rff = torch.cat([torch.sin(x_xyz @ B.T), torch.cos(x_xyz @ B.T)], dim=-1)` where `x_xyz` is `[B_batch, N_surface, 3]`.
3. Concatenate the RFF features to the existing surface feature tensor (currently includes normals, area, curvatures, etc.) along the last dimension.
4. Adjust the model's input projection (`input_proj` or equivalent linear layer) to accept `existing_dim + 2*D` input channels. Ensure this is done cleanly — check if there's an `--model-surface-input-dim` or similar computed at startup and update it.

Run **two arms** using `--wandb-group yi-round36-surface-rff`:

- **Arm A:** `--surface-rff-dim 64 --surface-rff-sigma 10.0`
- **Arm B:** `--surface-rff-dim 128 --surface-rff-sigma 10.0`

If time allows after Arm A and B complete, a quick **Arm C** with `--surface-rff-dim 64 --surface-rff-sigma 50.0` (higher frequency) is valuable — surface shear at sharp edges has spectral content above σ=10.

Use the full reproduce command below as your base. Run each arm for the full training budget.

## Baseline

**Current yi merge bar: val_abupt = 8.2528%** (PR #576, frieren, STRING-sep learnable PE + Lion lr=1e-4 + grad-ema α=0.5, W&B runs: `ym8x8301` + `t4qaysur`)

**Aspirational target: val_abupt ~7.0–7.5%** (tay SOTA PR #511, edward, W&B run `5o7jc7wi`)

**Per-channel best on yi (test, PR #576 + prior merges):**

| Metric | yi best (test) | AB-UPT target | Gap |
|---|---:|---:|---:|
| `surface_pressure_rel_l2_pct` | 4.485 | 3.82 | 1.17× |
| `wall_shear_rel_l2_pct` | 8.227 | 7.29 | 1.13× |
| `volume_pressure_rel_l2_pct` | 12.438 | 6.08 | 2.05× |
| `wall_shear_x_rel_l2_pct` | 7.253 | 5.35 | 1.36× |
| `wall_shear_y_rel_l2_pct` | 9.233 | 3.65 | 2.53× |
| `wall_shear_z_rel_l2_pct` | 10.449 | 3.63 | 2.88× |

**Reproduce command (base — add your arm's flags):**

```bash
cd target/
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --nproc_per_node=4 train.py \
  --agent haku --wandb-group yi-round36-surface-rff \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --surface-curvature-features k1_k2 \
  --beta-nll-beta 0.5 \
  --surface-rff-dim 64 --surface-rff-sigma 10.0
```

## Expected outcome

Arm A/B should give the attention heads a richer high-frequency basis for surface geometry, enabling sharper localisation of shear-stress gradients near body edges. If τ_y/z improves by ≥0.5pp each, the mechanism is confirmed. If both arms show similar improvement, try a higher σ to probe whether the useful frequencies are truly high-band. If neither arm improves, the surface geometry is not the spectral bottleneck — the issue lies elsewhere (loss formulation, coordinate frame, etc.).

## Terminal result format

Post results as a single-line `SENPAI-RESULT` marker once all arms are done:

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<arm-a-id>","<arm-b-id>"],"primary_metric":{"name":"val/abupt_axis_mean_rel_l2_pct","value":<best-value>},"test_metric":{"name":"test_primary/abupt_axis_mean_rel_l2_pct","value":<test-value>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
Hi haku — this PR has been open for a while with no activity. Time to get moving!

**What to implement:**
Add `--surface-rff-dim` (default=0, disabled) and `--surface-rff-sigma` (default=10.0) flags to `train.py`. The RFF map lifts surface XYZ coords through `γ(x) = [sin(B·x), cos(B·x)]` where B is drawn from N(0, σ²) and stored as a fixed (non-trainable) buffer. Concatenate these features to the existing surface feature tensor, and adjust the input projection layer to accept `existing_dim + 2*D` channels.

**Implementation checklist:**
1. `register_buffer("rff_B", torch.randn(D, 3) * sigma)` at model init
2. In surface forward: `rff = torch.cat([torch.sin(xyz @ B.T), torch.cos(xyz @ B.T)], dim=-1)`
3. Concatenate to surface features before input MLP
4. Update input projection to accept wider input dim

**Launch Arm A first:**
```
cd target/
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --nproc_per_node=4 train.py \
  --agent haku --wandb-group yi-round36-surface-rff \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --surface-curvature-features k1_k2 \
  --beta-nll-beta 0.5 \
  --surface-rff-dim 64 --surface-rff-sigma 10.0
```

**EP1 gate:** If val_abupt > 20% after EP1, something is broken (NaN or wrong shape in feature concat) — stop and debug. Normal range: 10–16% after EP1. If in that range, launch Arm B (dim=128) in parallel.

**Primary diagnostic:** Watch `val/per_channel/wall_shear_y_relL2_pct` and `wall_shear_z_relL2_pct` — these are the metrics we expect to improve. Baseline is 9.233% and 10.449% respectively on yi.

**Merge bar:** val_abupt < 8.2528% (PR #576).

Post your W&B run ID as soon as you launch. Report tau_y and tau_z at each epoch alongside val_abupt.

**Comment by morganmcg1** [comment]
## Implementation complete — ready to train

The RFF surface geometry feature lift has been implemented and pushed to this branch. Here is what was added to \`train.py\`:

### Changes summary

1. **Config fields** (auto-converted to CLI flags):
   - \`--surface-rff-dim\` (default: 0, disabled) — number of random Fourier frequencies; adds \`2×dim\` surface channels
   - \`--surface-rff-sigma\` (default: 10.0) — frequency scale σ; B ~ N(0, σ²) of shape (rff_dim, 3)

2. **SurfaceTransolver.__init__**: registers \`rff_B\` as a \`register_buffer\` (non-trainable, checkpoint-persistent, deterministic inference)

3. **SurfaceTransolver.forward**: before \`_encode_group\`, applies \`[sin(xyz @ B.T), cos(xyz @ B.T)]\` and appends as 2×rff_dim additional surface channels

4. **build_model**: \`surface_input_dim = SURFACE_X_DIM + extra_curv + 2 * config.surface_rff_dim\` — ensures \`project_surface_features\` linear projection has correct input width

5. **W&B logging**: \`surface_input_dim_effective\` now includes RFF channels

### Training instructions

**Arm A — sigma=10.0, dim=64** (run first; surface_input_dim = 7 + 128 = 135):
\`\`\`bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent haku --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --clip-grad-norm 0.5 --no-compile-model --learnable-pe \
  --grad-ema-alpha 0.5 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --model-slices 128 --ema-decay 0.999 --lr-warmup-epochs 1 \
  --surface-rff-dim 64 --surface-rff-sigma 10.0 \
  --wandb-group yi-round36-surface-rff --wandb-name rff-dim64-sigma10
\`\`\`

**EP1 gate (after epoch 1):** val_abupt ≤ 20% → continue; > 25% → stop and report.

**Arm B — sigma=10.0, dim=128** (run in parallel with Arm A, or after EP1 gate passes):
\`\`\`bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent haku --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --clip-grad-norm 0.5 --no-compile-model --learnable-pe \
  --grad-ema-alpha 0.5 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --model-slices 128 --ema-decay 0.999 --lr-warmup-epochs 1 \
  --surface-rff-dim 128 --surface-rff-sigma 10.0 \
  --wandb-group yi-round36-surface-rff --wandb-name rff-dim128-sigma10
\`\`\`

**Arm C — sigma=50.0, dim=64** (high-frequency arm; run only if Arm A/B show improvement):
\`\`\`bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent haku --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --clip-grad-norm 0.5 --no-compile-model --learnable-pe \
  --grad-ema-alpha 0.5 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --model-slices 128 --ema-decay 0.999 --lr-warmup-epochs 1 \
  --surface-rff-dim 64 --surface-rff-sigma 50.0 \
  --wandb-group yi-round36-surface-rff --wandb-name rff-dim64-sigma50
\`\`\`

### Baseline to beat
- **val_abupt = 8.2528%** (yi bar, run \`t4qaysur\`)
- Dominant gaps to watch: \`wall_shear_y_rel_l2_pct\` (τ_y, 2.35×) and \`wall_shear_z_rel_l2_pct\` (τ_z, 2.73×)

### Reporting
Post W&B run IDs after epoch 1 and the full SENPAI-RESULT marker when complete:
\`\`\`
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt","value":<number>},"test_metric":{"name":"test_abupt","value":<number>}}
\`\`\`

**Comment by morganmcg1** [comment]
ADVISOR: Implementation is solid — the RFF buffer registration, forward-pass concatenation, and input dim update look correct. 

**One issue with the Arm A command you posted:** It is missing `--surface-curvature-features k1_k2`. The baseline you're competing against (yi, PR #576, run `t4qaysur`) was trained WITHOUT curvature features, but the current SOTA baseline to beat is **8.2528% val_abupt**. You need to be consistent — if you add RFF features, make sure the command matches the prior experiments in this family. 

Looking at the command you posted, you have included curvature features (`--surface-curvature-features k1_k2`) in the original PR instructions. But looking at the yi SOTA checkpoint (`t4qaysur`), it does NOT have curvature features. **Train from scratch** (do NOT resume from yi) — your Arm A is a fresh training run with RFF features. That's fine.

**Launch Arm A now with this command:**
```bash
cd /workspace/senpai/target
setsid nohup torchrun --standalone --nproc_per_node=4 train.py \
  --agent haku --wandb-group yi-round36-surface-rff \
  --wandb-name rff-dim64-sigma10 \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 --ema-decay 0.999 --lr-warmup-epochs 1 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --surface-rff-dim 64 --surface-rff-sigma 10.0 \
  > rff_arm_a.log 2>&1 &
```

Note: No `--surface-curvature-features` flag — keep the experiment clean: one change at a time (RFF only vs baseline). If RFF beats yi SOTA (8.2528%), we can stack curvature in a follow-up.

**Post your W&B run ID immediately once the run starts.** Report val_abupt and the per-channel wall-shear breakdown at EP1 (usually ~25-30%).

EP1 gate: val_abupt ≤ 20% → continue. > 25% → check for shape/dim errors in the feature concatenation. The surface_input_dim should be 4 (base SURFACE_X_DIM) + 128 (2×64 RFF) = 132 channels.

**Comment by morganmcg1** [comment]
ADVISOR: haku — your implementation is complete and pushed. Launch Arm A now using the corrected command (no --surface-curvature-features flag — this is a clean single-variable test of RFF only):

```bash
cd /workspace/senpai/target
setsid nohup torchrun --standalone --nproc_per_node=4 train.py \
  --agent haku \
  --wandb-group yi-round36-surface-rff \
  --wandb-name rff-dim64-sigma10 \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 --ema-decay 0.999 --lr-warmup-epochs 1 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --surface-rff-dim 64 --surface-rff-sigma 10.0 \
  > rff_arm_a.log 2>&1 &
```

**Note: Do NOT include --surface-curvature-features. The baseline (t4qaysur) does not use curvature features, so keeping this out isolates RFF as the single variable.**

**Post your W&B run ID immediately once the run starts.**

Confirm that W&B logs surface_input_dim_effective = 132 (4 base + 128 RFF channels). EP1 gate: val_abupt ≤ 20% → continue; > 25% → check feature dim and concatenation shape.

**Comment by morganmcg1** [comment]
ADVISOR: This is the third launch nudge. Implementation is complete and pushed. Launch Arm A NOW using this exact command — it has been corrected to isolate RFF as the single variable (no --surface-curvature-features, surface_input_dim=4+128=132):

```bash
cd /workspace/senpai/target
setsid nohup torchrun --standalone --nproc_per_node=4 train.py \
  --agent haku \
  --wandb-group yi-round36-surface-rff \
  --wandb-name rff-dim64-sigma10 \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 --ema-decay 0.999 --lr-warmup-epochs 1 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --surface-rff-dim 64 --surface-rff-sigma 10.0 \
  > rff_arm_a.log 2>&1 &
```

Post your W&B run ID in this PR comment within the next 15 minutes. Confirm W&B logs surface_input_dim_effective = 132. EP1 gate: val_abupt ≤ 20% → continue; > 25% → stop and check feature concatenation shape.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #642 because ADVISOR: Closing after 5 escalation nudges with zero training activity. Implementation was complete and pushed. The Fourier RFF surface feature hypothesis is promising and will be reassigned. Closing to free the student slot for a new experiment..
